### ISSUE TITLE
Add docs for docker and docker-compose

### ISSUE THREAD

**Issue by MaybeThisIsRu** [hometown-fork/hometown#666]
### Pitch

I'll be moving my instance from a lonesome, dedicated VPS to one that hosts a bunch of stuff via docker soon. The lack of docs for getting started with a docker-based setup in the past motivated me to just throw money at the problem and spin up a brand new VPS entirely for the Hometown instance.

The Mastodon docs themselves [offer a detailed guide to installing and setting up](https://docs.joinmastodon.org/admin/prerequisites/) a Mastodon instance. As the upstream devs say, a bare-machine install does offer a lower barrier to setup (for them and for the users). I followed the Mastodon docs and set Hometown up the last time around.

### Motivation

My argument is that docker itself _can_ offer an even lower barrier when documented properly. I'll be outlining steps I take with my migration of sorts, but in the meantime I'm opening this issue/feature request as a placeholder for others using Hometown's compose file to run instances to pitch in with:

- an introduction/overview of the setup,
- a step-by-step guide for first-time setup,
- any tips and gotchas to make note of while setting things up, and,
- any limitations or pain points to be aware of when opting for the docker-based setup over a system-level install.

**Comment by MaybeThisIsRu** [comment]
Just haven't gotten around to this yet... There's a chance I won't, at all. If SmallCamp is able to cover even 50% of its cost from recurring donations (which _might_ happen?), I would have no motivation left to move from my system-level installation to a docker-based setup.

**Comment by hirusi** [comment]
Does Hometown require anything more than what [Mastodon says is sufficient for making backups](https://docs.joinmastodon.org/admin/backups/)?

**Comment by MaybeThisIsRu** [comment]
Does Hometown require anything more than what [Mastodon says is sufficient for making backups](https://docs.joinmastodon.org/admin/backups/)?

Additionally, if anyone has any idea how the public/system directory would be moved to S3-compatible storage, that would be very nice.

I'm also scratching out my thoughts and notes here in this comment.

## The big picture

- Moving from bare-metal Hometown 1.0.5 to docker-compose based Hometown 1.0.5
- Moving from self-hosted `public/system` to B2 (which has an S3 compatible API)

## Steps

So far, I'm just through with the first step, but the idea is to:

### Preparation

1. Make database, media, application config, and Redis backups. I've tested the backup process, but not the import process yet.
2. Upload the `public/system` files to a B2 bucket (~40 GB).

### Actual migration

1. On a different VPS, bring up the docker containers from this repo with `docker-compose up` and investigate what the state of things is at this point.
  - Point to the backup application config (`.env.production`).
  - Create a named volume for the DB and dump the backup using `psql`.
  - ??? Configure Mastodon to use S3-compatible storage. How does the database understand the new paths?
  - Copy the Redis dump into its named volume as-is.
2. Update the DNS to this VPS.

### Post migration

Take things from there -- does everything work?

**Comment by hughrun** [comment]
Just a quick note to say a *new* official guide is apparently being worked on. See https://github.com/tootsuite/documentation/issues/778

Not entirely clear why the original guide (which I used successfully when I knew not much about Docker) was vapourised. 

Note I haven't moved ausglam.space to HomeTown yet but intend to once it's aligned with Mastodon 3.3.0

**Comment by MaybeThisIsRu** [comment]
Oh, thanks for that heads up! I had no idea they were working on one.

I decided not to migrate my Hometown instance to docker in the end. But I'm open to doing so in the future and therefore have left this issue as open.

As for...

> Does Hometown require anything more than what Mastodon says is sufficient for making backups?

No.

**Comment by morganmay** [comment]
I've been thinking about setting up a Hometown instance and I'd like to use docker if possible, so any documentation would be appreciated. My initial installation attempts didn't work out, and I haven't gotten back to it yet.

**Comment by dariusk** [comment]
I am open to Docker-related documentation for Hometown but one of my rules for this project has been "Don't bother with Docker because I know from my day job experience that it will drive me crazy and burn me out." So if someone want to help write the docs, 100% I am interested in this I just don't have the capacity for it. I'll tag this issue `help wanted`.

**Comment by hughrun** [comment]
@dariusk I took the opportunity to extract myself from Docker hell when migrating ausglam.space to Hometown so I'm afraid I will not be of much use on this documentation project.

**Comment by MaybeThisIsRu** [comment]
Yeah I have no plans either to be docker-izing it anymore. Donations have really come through for SmallCamp and I no longer want to deal with docker hell when the bare metal setup is paid for.

**Comment by dariusk** [comment]
This is great to hear, @MaybeThisIsRu!

**Comment by nileshtrivedi** [comment]
Having a Docker Compose file for Hometown will allow it to be installable with one-click in self-hosted PaaS like [CapRover](https://caprover.com) so this still seems worth doing.

**Comment by nileshtrivedi** [comment]
For the time being, we should probably rename or remove the [docker-compose.yml](https://github.com/hometown-fork/hometown/blob/hometown-dev/docker-compose.yml) because it refers to Mastodon's official images `tootsuite/mastodon:v3.5.5` which of course won't have Hometown's changes.

**Comment by dariusk** [comment]
@nileshtrivedi Will simply removing the `docker-compose.yml` be good enough, you think? In terms of making it clear via code that we don't support docker (I will of course add to the docs too). Just want to make sure you couldn't still somehow try to get docker to run and end up in a half-working state that would result in bugs being filed here by confused admins.

**Comment by nileshtrivedi** [comment]
Yes, removing the compose file is a reasonable option. The current file is neither documented, nor hardened for security/scalability.

Mastodon Docker Comprose related PRs to watch: https://github.com/mastodon/mastodon/pull/21044 and https://github.com/mastodon/mastodon/pull/21165

**Comment by robsimmons** [comment]
An alternative would be to just remove the three lines `tootsuite/mastodon:v3.5.5` — I've gotten what I believe to be a successful script for deploying hometown with Linode's instructions, which used docker-compose, and all that was needed was the deletion of those three lines (though I also deleted all the db stuff so I could used a managed Postgres): https://docs.google.com/document/d/1pgVfDZ3q4obBkQAmLcb198dz8b3BH9RWQznzf6nivro/edit?usp=sharing

**Comment by nileshtrivedi** [comment]
I have raised a pull request over at Mastodon repo for enabling Docker for *development*: https://github.com/mastodon/mastodon/pull/22206

I don't want to attempt to refactor Docker files *for production* just yet because there are [many pending Docker refactoring/hardening related PRs](https://github.com/mastodon/mastodon/pull/16244#issuecomment-1336370181)

**Comment by robsimmons** [comment]
It definitely sounds like moving people away from docker for production is the right move, yeah.

**Comment by dariusk** [comment]
@nileshtrivedi This might be relevant too? Forgive me if it's not, my brain simply refuses to think about Docker if I am not getting paid a salary to do so https://github.com/mastodon/mastodon/pull/22057

**Comment by nileshtrivedi** [comment]
@dariusk Yes, it would be useful as hometown image would get published on Docker Hub as `hometown-fork/homwtown` , enabling people to deploy it on prod with just a docker compose file using default instructions, without having to build the image themselves.

Meanwhile, I [discovered](https://github.com/mastodon/mastodon/pull/22206#issuecomment-1345464733) that Mastodon (and hence Hometown) already have a DevContainers (i.e. Docker + VS Code) based workflow for local development that's actually quite nice. It's not ideal because it uses yet another variant of Dockerfile present in [.devcontainer](https://github.com/hometown-fork/hometown/tree/hometown-dev/.devcontainer) but it works.
